### PR TITLE
Unravel

### PR DESCRIPTION
**Pull request by nwinter**
Make game dev simulate on the main thread for performance.

Update to latest version of CreateJS from our 3-year-old version, now using StageGL (allowing us to relax some restrictions, like only being able to use one spritesheet per layer).

Fix click lag handling by monkeypatching CreateJS.

Fix a memory leak that was exacerbated by main thread simulation.

### CODE REVIEW

File: app/views/play/level/PlayLevelView.coffee [hunk 1]
```diff
@@ -53,6 +53,7 @@ WebSurfaceView = require './WebSurfaceView'
 SpellPaletteView = require './tome/SpellPaletteView'
 
 require 'lib/game-libraries'
+window.Box2D = require('exports-loader?Box2D!vendor/scripts/Box2dWeb-2.1.a.3')
```
Comment (nwinter) [inline app/views/play/level/PlayLevelView.coffee]: @phoenixeliot Is this a problem for webpacking? Needed it for when we simulate game dev levels on the main thread.

File: app/lib/Angel.coffee [hunk 2]
```diff
@@ -232,16 +232,20 @@ module.exports = class Angel extends CocoClass
       @hireWorker()
 
   abort: ->
-    return unless @worker and @running
+    return unless @running
     @say 'Aborting...'
     @running = false
+    @work?.world?.goalManager?.destroy()
+    # TODO: somehow clear out more references to this old world to avoid leaking if in synchronous mode
```
Comment (sderickson) [inline app/lib/Angel.coffee]: Does this TODO need to happen before we deploy?

File: app/lib/Angel.coffee [hunk 3]
```diff
@@ -296,51 +310,80 @@ module.exports = class Angel extends CocoClass
   simulateSync: (work) =>
     console?.profile? "World Generation #{(Math.random() * 1000).toFixed(0)}" if imitateIE9?
     work.t0 = now()
-    work.world = testWorld = new World work.userCodeMap
+    work.world = work.world = new World work.userCodeMap
```
Comment (sderickson) [inline app/lib/Angel.coffee]: Double assignment

File: app/lib/Angel.coffee [hunk 4]
```diff
@@ -296,51 +310,80 @@ module.exports = class Angel extends CocoClass
...
+      goalManager.worldGenerationWillBegin()
+      work.world.setGoalManager goalManager
+    @beginSimulationSync work
+
+  beginSimulationSync: (work) ->
+    work.t1 = now()
+    #Math.random = work.world.rand.randf  # so user code is predictable - TODO: will this work when we do it on the main thread?
+    #Aether.replaceBuiltin('Math', Math)
+    #replacedLoDash = _.runInContext(window)
+    #_[key] = replacedLoDash[key] for key, val of replacedLoDash
```
Comment (sderickson) [inline app/lib/Angel.coffee]: Are we not doing this work here? If we aren't, this commented out code should be removed.

File: app/lib/Angel.coffee [hunk 5]
```diff
@@ -296,51 +310,80 @@ module.exports = class Angel extends CocoClass
...
+        progress = Math.min(progress, 0.9) if world.indefiniteLength
+        @publishGodEvent 'world-load-progress-changed', progress: progress  # Debounce? Need to publish at all?
+        @streamFrameSync work
+        if world.indefiniteLength and world.victory?
+          world.indefiniteLength = false
+        continuing = world.shouldContinueLoading simulationLoopStartTime, (->), false, (=> @simulateFramesSync(work) unless @destroyed)
+        return unless continuing
+      if world.indefiniteLength and i is world.totalFrames - 1
+        ++world.totalFrames
+      frame = world.getFrame i++  # TODO: handle non-user-code errors?
```
Comment (sderickson) [inline app/lib/Angel.coffee]: Have you tested this TODO?

File: app/lib/surface/LayerAdapter.coffee [hunk 6]
```diff
@@ -275,20 +275,23 @@ module.exports = LayerAdapter = class LayerAdapter extends CocoClass
...
+        return @_renderNewSpriteSheet(async)
       @onBuildSpriteSheetComplete({async:async}, builder)
       return sheet
 
   onBuildSpriteSheetComplete: (e, builder) ->
     return if @initializing or @destroyed
     @asyncBuilder = null
 
-    if builder.spriteSheet._images.length > 1
+    if false  # builder.spriteSheet._images.length > 1  # Don't need to do this with StageGL any more
```
Comment (sderickson) [inline app/lib/surface/LayerAdapter.coffee]: This code block can just be deleted

File: app/lib/surface/Mark.coffee [hunk 7]
```diff
@@ -219,7 +219,7 @@ module.exports = class Mark extends CocoClass
     shapeName = if @lank.thang.shape in ['ellipsoid', 'disc'] then 'ellipse' else 'rect'
     key = "#{shapeName}-debug-#{@lank.thang.collisionCategory}"
     DEBUG_SIZE = 10
-    unless key in @layer.spriteSheet.getAnimations()
+    unless key in @layer.spriteSheet.animations
```
Comment (sderickson) [inline app/lib/surface/Mark.coffee]: Are all these changes from `getAnimations()` to `animations` because this function is costly?

Comment (nwinter) [reply app/lib/surface/Mark.coffee]: The `get` style are deprecated and throw warnings in CreateJS 1.0, recommended to just use accessors directly.

**Pull request merged**